ChainSafe/lodestar
# Separate fork choice

Current Status:
- able to run the chain from genesis and restore from checkpoints
- tests untouched and likely failing

Big items:

- `StateContextCheckpointCache` has been internally refactored, and its public interface extended
  - add `pruneFinalized(cp: Checkpoint)` - will prune checkpoints from epochs before the finalized epoch
  - add `prune(finalized: Checkpoint, justified: Checkpoint)` - similar to `stateCache.prune` - if there are more epochs present than `MAX_EPOCHS`, will prune starting from the lowest epoch to the highest epoch, skipping the finalized and justified epochs.
  - add `getLatest(cp: Checkpoint): ITreeStateContext` - starting from `cp.epoch` and going backwards, searches the cache for hits with `cp.root`. "Finds the latest state context at the checkpoint where the checkpoint's root is cp.root" and the checkpoint's epoch is lte cp.epoch."
- `StateRegenerator` has been added - `src/chain/regen.ts`
  - it can regenerate states from blocks _that have already been processed_ , given `forkChoice`, `stateCache`, `checkpointStateCache`, and `db.blocks`
  - it intelligently attempts to perform the smallest amount of regeneration necessary to regen the requested state. To that end, it exposes three different methods.
  - it reuses some block processing machinery (`emitCheckpointEvent`, `runStateTransition`) so as to enable re-seeding the checkpoint cache
  - it throws a `RegenError` with well-defined error codes and types, this allows us to tighten the interface to disallow nulls.
  - it has one major TODO, it must queue requests instead of allowing multiple concurrent requests
  - `getPreState(block: BeaconBlock): Promise<ITreeStateContext>` - the same as `getPreState` that we were using in our block processor. It can request a more recent checkpoint state if the block occurs after many skip-slots, skipping epoch transitions :) 
  - `getCheckpointState(cp: Checkpoint): Promise<ITreeStateContext>` - returns the checkpoint state. It minimizes regeneration by first searching the checkpoint cache for checkpoint states with cp.root, which may then just be `processSlots`ed. Otherwise regens the 'ol-fashioned way', by state root.
  - `getState(stateRoot: Root): Promise<ITreeStateContext>` - returns the state whose root is stateRoot. This is where the actual block replay happens. It iterates backwards thru the block tree from the requested state, to the first state that we have cached, (searching the checkpointStateCache as well!), then replaying the blocks that were iterated thru to the requested state, once a state has been found.
- Adds an optional `EpochProcess` object to `EpochContext`.
  - This object is used in epoch transitions to simplify validator balance counting. It holds flattened validators, validator statuses (whether they can attest, whether they've attested for important recent states, etc), total attested balances, etc.
  - This ends up being very useful for updating the fork choice balances.
  - We will find more uses for this after this PR, likely will be very nice to scrape for metrics re network participation.
- Block processing refactored slightly
  - `getPreState` is now handled by the `StateRegenerator`
  - `runStateTransition` and other utilities moved to `src/chain/blocks/stateTransition.ts`
- Block/state archival tasks refactored
  - they are now both triggered by the `forkChoice:finalized` event, the `forkChoice:pruned` event is no more.
  - we also prune state caches, triggered by the `checkpoint` event, but not in a task as its a one-liner
  - block archival no longer loads _all_ blocks in `block` repository into memory at once, instead streams
  - block archival strategy, run before fork choice pruning
    - stream all blocks in `block` repository, one by one
    - if the block is more recent than the finalized block, ignore
    - if the finalized block is a descendant of the block, store in the blockArchive
    - delete all un-ignored blocks
  - state archival now archives the finalized state from `checkpointStateCache` (these will be candidates for weak subjectivity :) )
- Add `@chainsafe/lodestar-fork-choice` package
  - split into two 'levels', protoArray, and forkChoice
  - protoArray - the block dag, not much else
  - forkChoice
    - provides an interface very close to the spec
    - includes an `IForkChoiceStore`, very similar to `Store` in the spec -- this is implemented in the lodestar package with knowledge of the `ChainEventEmitter`
    - stores votes, balances, queued attestations
  - Attestation weighting is completely different than before
    - Instead of feeding attestations in with weights attached per attestation, we feed in something like a `LatestMessage` from the spec.
    - attestations are reweighted all-at-once, using `forkChoice.updateBalances`. Using `epochContext.epochProcess`, this is fast, lots of the work was already done as part of epoch processing.
    - attestation reweighting is triggered by the `forkChoice:justified` event.
